core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -1992,4 +1992,32 @@ package object config {
       .version("3.1.0")
       .doubleConf
       .createWithDefault(5)
+
+  private[spark] val SHUFFLE_NUM_PUSH_THREADS =
+    ConfigBuilder("spark.shuffle.push.numPushThreads")
+      .doc("Specify the number of threads in the block pusher pool. These threads assist " +
+        "in creating connections and pushing blocks to remote shuffle services when push based " +
+        "shuffle is enabled. By default, the threadpool size is equal to the number of cores.")
+      .version("3.1.0")
+      .intConf
+      .createOptional
+
+  private[spark] val SHUFFLE_MAX_BLOCK_SIZE_TO_PUSH =
+    ConfigBuilder("spark.shuffle.push.maxBlockSizeToPush")
+      .doc("The max size of an individual block to push to the remote shuffle services when push " +
+        "based shuffle is enabled. Blocks larger than this threshold are not pushed.")
+      .version("3.1.0")
+      .bytesConf(ByteUnit.KiB)
+      .createWithDefaultString("800k")
```

Comment:
It looks like a magic number.
- Is `800k` optimal at some kind of workload?
- If we need to choose around 800KiB, can we use `1024k (1MiB)` instead?

Reply:
@Victsm could you please address this one?

Reply:
Agree that 1MB as the default value is perhaps more appropriate.
We are right now achieving 2 things with this configuration:

1. It serves as a threshold to determine whether a block is large enough to not trigger a potential disk throughput issue.
2. It also serves as a static threshold to determine if a block is potentially coming from a skewed partition, which would likely be much larger than the other blocks generated by the current mapper.

For the 1st need, a static configuration is perhaps sufficient.
For the 2nd need, not sure if we should also look at the distribution of the block sizes for a given mapper, which is known to the ShuffleBlockPusher at the beginning of `initiateBlockPush`.
If the total size of the skewed partition is not way too large (a few GB instead of 10s or 100s GB), it's probably still fine to merge it.
If the skewed partition size is too large, merging it into a single shuffle service is probably not ideal with the possibility of overloading that shuffle service.

This is also where push-based shuffle and AQE skew join intersects.
Since the blocks get merged into a merged shuffle partition in random order, that merged shuffle partition cannot be further divided for AQE skew join.
Since push-based shuffle maintains both the merged shuffle files and the original shuffle files, we currently bypass skewed partitions in push-based shuffle and let AQE handle these partitions.
This is part of #30480 
For AQE skewed join, it uses a skew factor to determine which partitions are skewed.
I'm wondering whether we should take similar approaches here to decide which blocks should be considered as potentially coming from a skewed partition.

@tgravescs @attilapiros @Ngone51 @jiangxb1987 @mridulm @dongjoon-hyun want to know your thoughts on this as well.

Reply:
@Victsm 
For 1, I think 1MiB is better(that's what Spark usually do) if you do not see much performance difference between 800Kib and 1 MiB. And you'd better add more explanation in the conf doc to say something like the default value is the appropriate value to avoid potential disk throughput issue and a small value could lead to severe disk issue.

For 2, AFAIK, Spark doesn't know the details of AQE at the shuffle level yet. So we actually don't even know whether there's a join operation at SQL query. So how can we decide whether the skewed partition needs to be calculated or not inside ShuffleBlockPusher? Besides, if we want to calculate the skewed partition, we should ensure it's the same(or less than) as the calculated one at AQE level, right? Could we make sure of it? (Maybe this's easy if two places use the same algorithm)

It's might possible when AQE can pass more info through the task and ShuffleBlockPusher leverage it then. But for now, I feel it's kind of hard to intersect with AQE.

Reply:
I will change the default to 1MiB. It seems that everyone is in agreement that it is the better default value.
cc @dongjoon-hyun @Victsm @Ngone51

Reply:
@Ngone51 agree that we cannot have perfect alignment between AQE and push-based shuffle right now, due to these 2 operating at 2 different levels.
The current implementation is rather an opportunistic approach in the hope that large blocks possibly also come from skewed partitions which if the shuffle is for a join and AQE is enabled, the corresponding partition could be caught as a skewed partition by AQE.
If this happens, the skewed partition would be consumed differently by AQE, not leveraging the merged shuffle partitions by push-based shuffle. Thus, pushing these blocks could be wasteful.
With the current limitations of AQE, i.e. skew handling is only for join and it's still largely based on the original shuffle mechanism, it is not straightforward to intersect push-based shuffle with AQE.
This could also be an area for further improvements down the way.